File: src/unsupported/thickset/test/test_ThickInterval.cpp
#define CATCH_CONFIG_MAIN
#include <catch.hpp>
#include "../ThickInterval.h"
#include <sstream>
// using ibex::ThickInterval;
using ibex::Interval;

TEST_CASE("Test Constructor", ""){
	SECTION(""){
		ThickInterval a(Interval(1, 2), Interval(2,3));
		CHECK(a.lb() == Interval(1, 2));
		CHECK(a.ub() == Interval(2, 3));
	}

	SECTION(""){
		ThickInterval a(1,2);
		CHECK(a.lb() == Interval(1, 1));
		CHECK(a.ub() == Interval(2, 2));
	}

	SECTION(""){
		ThickInterval a(Interval(1,2));
		CHECK(a.lb() == Interval(1, 1));
		CHECK(a.ub() == Interval(2, 2));
	}

	SECTION(""){
		ThickInterval a(Interval(0, 3), Interval::POS_REALS);
		CHECK(a.lb() == Interval(0, 3));
		CHECK(a.ub() == Interval::POS_REALS);
	}

	SECTION(""){
		ThickInterval a(Interval::NEG_REALS, Interval(-4, 0));
		CHECK(a.lb() == Interval::NEG_REALS);
		CHECK(a.ub() == Interval(-4, 0));
	}

	SECTION(""){
		ThickInterval a(Interval::NEG_REALS, Interval(-4, 0));
		CHECK(a.lb() == Interval::NEG_REALS);
		CHECK(a.ub() == Interval(-4, 0));
	}

	SECTION(""){
		ThickInterval a(Interval::ALL_REALS, Interval::ALL_REALS);
		CHECK(a.lb() == Interval::ALL_REALS);
		CHECK(a.ub() == Interval::ALL_REALS);
	}

	SECTION(""){
		ThickInterval a(Interval::EMPTY_SET, Interval::EMPTY_SET);
		CHECK( a.is_empty() );
	}


	// SECTION("Cst constructor"){
	// 	ThickInterval res(Interval::EMPTY_SET, Interval::EMPTY_SET);
	// 	CHECK(ThickInterval::EMPTY_THICK == res);
	// }
}

TEST_CASE("Test Invalid constructor", ""){
	SECTION(""){
		ThickInterval a(Interval(2,-2));
		CHECK( a.is_empty() );
	}

	SECTION(""){
		ThickInterval a(Interval(2,3), Interval(1, 2));
		CHECK( a.is_empty() );
	}

	SECTION(""){
		ThickInterval a(Interval(1,3), Interval(1, 2));
		CHECK( a.is_empty() );
	}

	SECTION(""){
		ThickInterval a(Interval(1,2), Interval(1, 2));
		CHECK(a.lb() == Interval(1,2));
		CHECK(a.ub() == Interval(1,2));
	}


	SECTION(""){
		ThickInterval a(Interval::POS_REALS, Interval::NEG_REALS);
		CHECK(a.is_empty() );
		// CHECK(a.ub() == Interval::EMPTY_SET);
	}
}


TEST_CASE("Intersects"){

	ThickInterval a(Interval(-2, 1), Interval(0, 3));
	ThickInterval b(Interval(-2, 1), Interval(2, 3));
	ThickInterval c(Interval(1, 2), Interval(1, 2));
  CHECK( a.intersects( Interval(0, 1) ) == true );
  CHECK( a.intersects( Interval(1,2) )  == false);
  CHECK( b.intersects( Interval(1,2) )  == true);
  CHECK( c.intersects( Interval::EMPTY_SET) == false);
  CHECK( c.intersects( Interval(1,2) ) ==true );

  CHECK( a.intersects( ThickInterval( Interval(-0.0001, 0.5),Interval(2, 3) ) ) == false );
	CHECK( a.intersects( ThickInterval( Interval(-1, 0), Interval(2, 3) ) )  == true );

}

TEST_CASE("test is degenerated"){

	CHECK( ThickInterval(Interval(-2,-2),Interval(-1,3)).is_degenerated() == true);
	CHECK( ThickInterval(Interval(-2,-2),Interval(-1,3)).is_interval()  == false);

	CHECK( ThickInterval(Interval(-3,-1),Interval(4,4)).is_degenerated()  == true);
	CHECK( ThickInterval(Interval(-3,-1),Interval(4,4)).is_interval()  == false);

	CHECK( ThickInterval(Interval(-1,-1),Interval(4,4)).is_degenerated()  == true);
	CHECK( ThickInterval(Interval(-1,-1),Interval(4,4)).is_interval()  == true);

}


TEST_CASE(" test supset "){
	CHECK( ThickInterval(Interval(-2, 1), Interval(0, 4)).superset() == Interval(-2, 4) );
	// empty set
	CHECK( ThickInterval(Interval(-2, 1), Interval(5, 4)).superset() == Interval::EMPTY_SET );
}

TEST_CASE(" test subset "){
  CHECK( ThickInterval(Interval(-2, 1), Interval(0, 4)).subset() == Interval::EMPTY_SET );
  CHECK( ThickInterval(Interval(-2, 1), Interval(4, 5)).subset() == Interval(1, 4) );
}

TEST_CASE(" test Intersection "){
	ThickInterval a( Interval(0, 2), Interval(4, 5));
	ThickInterval b( Interval(1,2), Interval(4, 10));

	SECTION(" intersection 1 "){
			CHECK( a.inter(b) == ThickInterval(Interval(1,2), Interval(4, 5)) );
	}
	SECTION(" interseciont 2 "){
		CHECK( (a & b) ==  ThickInterval(Interval(1, 2), Interval(4, 5)) );
	}
	// SECTION(" intersection 3 "){
		// ThickInterval c( Interval(-2,-1), Interval(3, 4	));
		// CHECK( (a & c) == ThickInterval(ibex::Interval::EMPTY_SET,ibex::Interval::EMPTY_SET) );
	// }
}

File: prev_versions/v1.5.2/python/codac/tests/tests_from_pyibex/test_thickInterval.py
#!/usr/bin/env python

# Codac tests
# Originated from the former pyIbex library (Benoît Desrochers)
# ---------------------------------------------------------------------------
# \date      2022
# \author    Benoît Desrochers
# \copyright Copyright 2022 Codac Team
# \license   This program is distributed under the terms of
#            the GNU Lesser General Public License (LGPL).

import unittest
from codac import *
from codac.unsupported import *


class TestThickInterval(unittest.TestCase):

  def check_thickinterval(self, y, res_lb, res_ub):
    self.assertEqual(y.lb(), res_lb)
    self.assertEqual(y.ub(), res_ub)

  def test_constructor_01(self):
    a = ThickInterval([1,2], [2,3])
    self.check_thickinterval(a, Interval(1,2), Interval(2,3))

  def test_constructor_02(self):
    a = ThickInterval(1,2)
    self.check_thickinterval(a, Interval(1), Interval(2))

  def test_constructor_03(self):
    a = ThickInterval(Interval(0, 3), Interval.POS_REALS);
    self.assertEqual(a.lb() , Interval(0, 3));
    self.assertEqual(a.ub() , Interval.POS_REALS);

  def test_constructor_04(self):
    a = ThickInterval(Interval.NEG_REALS, Interval(-4, 0));
    self.assertEqual(a.lb() , Interval.NEG_REALS);
    self.assertEqual(a.ub() , Interval(-4, 0));

  def test_constructor_05(self):
    a = ThickInterval(Interval.NEG_REALS, Interval(-4, 0));
    self.assertEqual(a.lb() , Interval.NEG_REALS);
    self.assertEqual(a.ub() , Interval(-4, 0));

  def test_constructor_06(self):
    a = ThickInterval(Interval.ALL_REALS, Interval.ALL_REALS);
    self.assertEqual(a.lb() , Interval.ALL_REALS);
    self.assertEqual(a.ub() , Interval.ALL_REALS);

  def test_constructor_07(self):
    a = ThickInterval(Interval.EMPTY_SET, Interval.EMPTY_SET);
    self.assertTrue( a.is_empty() );

  def test_invalid_constructor_01(self):
    a = ThickInterval(Interval(2,-2));
    self.assertTrue( a.is_empty() );

  def test_invalid_constructor_02(self):
    a = ThickInterval(Interval(2,3), Interval(1, 2));
    self.assertTrue( a.is_empty() );

  def test_invalid_constructor_03(self):
    a = ThickInterval(Interval(1,3), Interval(1, 2));
    self.assertTrue( a.is_empty() );

  def test_invalid_constructor_04(self):
    a = ThickInterval(Interval(1,2), Interval(1, 2));
    self.assertEqual(a.lb(), Interval(1,2));
    self.assertEqual(a.ub(), Interval(1,2));

  def test_invalid_constructor_01(self):
    a = ThickInterval(Interval.POS_REALS, Interval.NEG_REALS);
    self.assertTrue(a.is_empty());

  def test_intersect(self):
    a = ThickInterval(Interval(-2, 1), Interval(0, 3));
    b = ThickInterval(Interval(-2, 1), Interval(2, 3));
    c = ThickInterval(Interval(1, 2), Interval(1, 2));
    self.assertTrue(a.intersects(Interval(0, 1)) == True );
    self.assertTrue(a.intersects(Interval(1,2))  == False);
    self.assertTrue(b.intersects(Interval(1,2))  == True);
    self.assertTrue(c.intersects(Interval.EMPTY_SET) == False);
    self.assertTrue(c.intersects(Interval(1,2)) ==True );
    self.assertTrue(a.intersects(ThickInterval(Interval(-0.0001, 0.5),Interval(2, 3))) == False);
    self.assertTrue(a.intersects(ThickInterval(Interval(-1, 0), Interval(2, 3)))  == True);

  def test_is_degenerated(self):
    self.assertTrue(ThickInterval(Interval(-2,-2),Interval(-1,3)).is_degenerated());
    self.assertFalse(ThickInterval(Interval(-2,-2),Interval(-1,3)).is_interval());
    self.assertTrue(ThickInterval(Interval(-3,-1),Interval(4,4)).is_degenerated());
    self.assertFalse(ThickInterval(Interval(-3,-1),Interval(4,4)).is_interval());
    self.assertTrue(ThickInterval(Interval(-1,-1),Interval(4,4)).is_degenerated());
    self.assertTrue(ThickInterval(Interval(-1,-1),Interval(4,4)).is_interval());

  def test_superset(self):
    self.assertEqual(ThickInterval(Interval(-2, 1), Interval(0, 4)).superset(), Interval(-2, 4));
    self.assertEqual(ThickInterval(Interval(-2, 1), Interval(5, 4)).superset(), Interval.EMPTY_SET);

  def test_subset(self):
    self.assertEqual(ThickInterval(Interval(-2, 1), Interval(0, 4)).subset(), Interval.EMPTY_SET);
    self.assertEqual(ThickInterval(Interval(-2, 1), Interval(4, 5)).subset(), Interval(1, 4) );


class TestIntersectThickInterval(unittest.TestCase):

  def setUp(self):
    self.a = ThickInterval( Interval(0, 2), Interval(4, 5));
    self.b = ThickInterval( Interval(1,2), Interval(4, 10));

  def test_intersect_01(self):
    res = ThickInterval([1,2],  [4, 5])
    self.assertEqual(self.a.inter(self.b), res)

  def test_intersect_02(self):
    res = ThickInterval([1,2],  [4, 5])
    self.assertEqual((self.a & self.b), res)


class TestisSubsetThickInterval(unittest.TestCase):

  def setUp(self):
    self.a = ThickInterval(Interval(0, 2), Interval(4, 5));
    self.b = ThickInterval(Interval(1,2), Interval(4, 10));

  def test_isSubset_01(self):
    res = self.a.is_inside(Interval(-1, 6))
    self.assertEqual(res, IN)
    res = self.a.is_inside(Interval(0, 5))
    self.assertEqual(res, IN)

  def test_isSubset_02(self):
    res = self.a.is_inside(Interval(-4, -1))
    self.assertEqual(res, OUT)

  def test_isSubset_03(self):
    res = self.a.is_inside(Interval(3, 5))
    self.assertEqual(res, MAYBE)

  def test_isSubset_04(self):
    res = self.a.is_inside(Interval(1, 5))
    self.assertEqual(res, MAYBE_IN)

  def test_isSubset_05(self):
    res = self.a.is_inside(Interval(-1, 1))
    self.assertEqual(res, MAYBE_OUT)
    res = self.a.is_inside(Interval(-1, 0))
    self.assertEqual(res, MAYBE_OUT)


if __name__ == '__main__':

  unittest.main()